### PR TITLE
Man pages

### LINKED ISSUE #5605: Feature: generate man pages
It'd be really nice to be able to generate man page automatically. The [flags library](https://pkg.go.dev/github.com/jessevdk/go-flags) lnd is using already [supports generating manpage](https://pkg.go.dev/github.com/jessevdk/go-flags#Parser.WriteManPage) although a bit awkwardly - by using same code that is used in runtime.

**Possible approaches**

* Have an additional switch/command in `lncli` (and ideally also in `lnd`) that will save the man page in a file/stdout. This is somewhat weird in cross-compiling scenarios.
* Have a very small binary that imports same config code and that binary writes manpage.
* Separate binary as before + automatically run it in release script (never cross compile) and add manpage to the release - I like this the most.

### PR DESCRIPTION
**Pull request by mauricepoirrier**
Paired with @lightningorb

## Change Description
This commits aim to generate man pages docs of lnd and lncli using formatting tools from already used packages.

To do this we:
1. add a flag on lnd that given that generates documentation and exits
2. add a hidden Command to lncli that generates documentation
3. add a makefile command to generate both docs depending on the installation / building

Haven't checked tests in these files.

closes #5605

## Steps to Test
```bash
make build
make docs
```
or 
```bash
make build
./lncli-debug generatemanpage
./lnd-debug --manpage
```

### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

### CODE REVIEW

**Review by ellemouton — CHANGES_REQUESTED** [review]
Hi @mauricepoirrier and @lightningorb, Super cool feature!! 🚀 

A few comments/suggestions from me: 
- At the moment, the man pages have to be manually moved to `usr/local/shared/man/man1` in order for the `man` command to work. I think this should be automated.
- Currently the `lnd-debug` and `lncli-debug` binaries are used to gen the man files which is not ideal since those are the names that appear in the man file. 
- Building on the above, I think `make install` should take care of everything: it should install lnd & lncli, build the man pages for both, and move the man pages to the correct locations. 
- the commit titles look good but the commit messages need to pls be [wrapped correctly](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md)
- The PR is also missioning a release notes entry

File: Makefile [hunk 1]
```diff
@@ -307,6 +307,16 @@ clean-mobile:
 	$(RM) -r mobile/build
 	$(RM) mobile/*_generated.go
 
+docs:
+	@$(call print, "Generating man pages documents.")
+	if [ -f ./lnd-debug ] && [ -f ./lncli-debug ]; then \
+	./lnd-debug --manpage; \
```
Comment (guggero) [inline Makefile:313]: Can we just depend on the `build` step to make sure the `lnd-debug` and `lncli-debug` files are present?

File: cmd/lncli/commands.go [hunk 2]
```diff
@@ -1323,6 +1323,30 @@ func channelBalance(ctx *cli.Context) error {
...
 
+var generateManPageCommand = cli.Command{
+	Name:   "generatemanpage",
+	Usage:  "Generates a man page for lncli.",
+	Hidden: true,
+	Action: actionDecorator(generateManPage),
+}
+
+func generateManPage(ctx *cli.Context) error {
+	docs, manErr := ctx.App.ToMan()
```
Comment (guggero) [inline cmd/lncli/commands.go:1334]: Can just create and re-use a single `err` variable, it's easier to read if every error has the same name.

File: cmd/lncli/commands.go [hunk 3]
```diff
@@ -1323,6 +1323,30 @@ func channelBalance(ctx *cli.Context) error {
...
+        if manErr != nil {
+		return manErr
+        }
+
+	writeErr := ioutil.WriteFile(
+		"lncli.1",
```
Comment (guggero) [inline cmd/lncli/commands.go:1340]: Can use multiple arguments per line, up to the 80 character limit.

File: config.go [hunk 4]
```diff
@@ -686,6 +688,17 @@ func LoadConfig(interceptor signal.Interceptor) (*Config, error) {
 		os.Exit(0)
 	}
 
+	if preCfg.GenerateManPage {
+		fileParser := flags.NewParser(&preCfg, flags.Default)
```
Comment (guggero) [inline config.go:692]: Couldn't this also be done in the `lncli generatemanpage` command? Doesn't look like we need the actual content of the `preCfg`, so we can just use the default config.

File: config.go [hunk 5]
```diff
@@ -686,6 +688,17 @@ func LoadConfig(interceptor signal.Interceptor) (*Config, error) {
...
+		var buf bytes.Buffer
+		fileParser.WriteManPage(&buf)
+		err := ioutil.WriteFile("lnd.1", buf.Bytes(), 0644)
```
Comment (guggero) [inline config.go:695]: Should the generated files be added to the `.gitignore` file?

Comment (guggero) [inline config.go:695]: Should use `os.WriteFile` instead of the deprecated `ioutil` one.

**Review by guggero — COMMENTED** [review]
Thanks for the PR. This should be quite useful for package maintainers.

Do you know I tell the `man` command to correctly show the generated files? It looks scrambled if I do `man -l lnd` for example. Probably just a format/encoding issue.

File: config.go (hunk 4 above)
Comment (mohamedawnallah) [reply config.go:692]: Hey @guggero ! could you please elaborate more on this?

**Comment by mohamedawnallah** [comment]
> * At the moment, the man pages have to be manually moved to `usr/local/shared/man/man1` in order for the `man` command to work. I think this should be automated.
> * Building on the above, I think `make install` should take care of everything: it should install lnd & lncli, build the man pages for both, and move the man pages to the correct locations.

Hey @ellemouton moving the man pages of `lnd` and `lncli` to `/usr/local/share/man/man1` would require the user to enter the password while running `make install` since the owner of this directory is the `root` ?

**Comment by mohamedawnallah** [comment]
I'm interested in taking on this issue. All seems good to me I've just put the comments where I have some doubts about when implementing this feature. Thanks!

**Comment by Kixunil** [comment]
@mohamedawnallah the canonical way to resolve these things is to write use `$DESTDIR` and `$PREFIX` variables. E.g.:

```make
install:
    install -m 755 lnd $(DESTDIR)$(PREFIX)/bin/lnd # same with lncli etc
    install -m 644 lnd.1 $(DESTDIR)$(PREFIX)/share/man/man1/lnd.1
    install -m 644 lncli.1 $(DESTDIR)$(PREFIX)/share/man/man1/lncli.1
```

This takes care of several issues:

* Non-packaging users usually want `PREFIX` to be `/usr/local` while packaging users want it to be `/`
* Packagers want to actually install into a temporary directory which can be set by `DESTDIR`, this is just empty in non-packaging scenarios
* Yes, users who want to run `make install` directly have to run it as root, this is expected.

Note that `PREFIX` is usually configured with a `./configure` script with `--prefix=` parameter. Though there's probably a simpler solution since LND doesn't need to know the prefix.

File: config.go (hunk 4 above)
Comment (mohamedawnallah) [reply config.go:692]: Thanks, I figured it out while working on the coin-selection-strategy option issue. I just needed more time to get hands-on with the codebase!

**Pull request closed without merging**